p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,258 @@
+// and by remote address from which the connection has received valid STUN/RTC
+// packets.
+//
+// When a new packet is received on the underlying net.PacketConn, we
+// first check the address map to see if there is a connection associated with the
+// remote address. If found we forward the packet to the connection. If an associated
+// connection is not found, we check to see if the packet is a STUN packet. We then
+// fetch the ufrag of the remote from the STUN packet and use it to check if there
+// is a connection associated with the (ufrag, IP address family) pair. If found
+// we add the association to the address map. If not found, it is a previously
+// unseen IP address and the `unknownUfragCallback` callback is invoked.
+type udpMux struct {
+	socket               net.PacketConn
+	unknownUfragCallback func(string, net.Addr)
+
+	m        sync.Mutex
+	ufragMap map[ufragConnKey]*muxedConnection
+	addrMap  map[string]*muxedConnection
+
+	// the context controls the lifecycle of the mux
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewUDPMux(socket net.PacketConn, unknownUfragCallback func(string, net.Addr)) ice.UDPMux {
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := &udpMux{
+		ctx:                  ctx,
+		cancel:               cancel,
+		socket:               socket,
+		unknownUfragCallback: unknownUfragCallback,
+		ufragMap:             make(map[ufragConnKey]*muxedConnection),
+		addrMap:              make(map[string]*muxedConnection),
+	}
+
+	mux.wg.Add(1)
+	go mux.readLoop()
+	return mux
+}
+
+// GetListenAddresses implements ice.UDPMux
+func (mux *udpMux) GetListenAddresses() []net.Addr {
+	return []net.Addr{mux.socket.LocalAddr()}
+}
+
+// GetConn implements ice.UDPMux
+// It creates a net.PacketConn for a given ufrag if an existing
+// one cannot be  found. We differentiate IPv4 and IPv6 addresses
+// as a remote is capable of being reachable through multiple different
+// UDP addresses of the same IP address family (eg. Server-reflexive addresses
+// and peer-reflexive addresses).
+func (mux *udpMux) GetConn(ufrag string, addr net.Addr) (net.PacketConn, error) {
+	a, ok := addr.(*net.UDPAddr)
+	isIPv6 := ok && a.IP.To4() == nil
+	return mux.getOrCreateConn(ufrag, isIPv6)
+}
+
+// Close implements ice.UDPMux
+func (mux *udpMux) Close() error {
+	select {
+	case <-mux.ctx.Done():
+		return nil
+	default:
+	}
+	mux.cancel()
+	mux.socket.Close()
+	mux.wg.Wait()
+	return nil
+}
+
+// RemoveConnByUfrag implements ice.UDPMux
+func (mux *udpMux) RemoveConnByUfrag(ufrag string) {
+	if ufrag == "" {
+		return
+	}
+	mux.m.Lock()
+	defer mux.m.Unlock()
+	for _, isIPv6 := range []bool{true, false} {
+		key := ufragConnKey{ufrag: ufrag, isIPv6: isIPv6}
+		if conn, ok := mux.ufragMap[key]; ok {
+			_ = conn.closeConnection()
+			delete(mux.ufragMap, key)
+			for _, addr := range conn.addresses {
+				// log.Errorf("deleting address : %v %v", ufrag, addr)
```

Comment:
commented out line can be deleted?

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)